plugins/enthusiast-common/enthusiast_common/agents/base.py

```diff
@@ -0,0 +1,72 @@
+from abc import ABC, abstractmethod
+from typing import Sequence, Self
+
+from langchain.agents import AgentExecutor, create_tool_calling_agent
+from langchain.memory import ConversationSummaryBufferMemory
+from langchain_core.language_models import BaseLanguageModel
+from langchain_core.prompts import PromptTemplate, ChatPromptTemplate
+from langchain_core.prompts.chat import MessageLikeRepresentation
+from langchain_core.tools import BaseTool
+
+from ..services import ConversationService
+
+
+class BaseAgent(ABC):
+    _agent_executor: AgentExecutor
+
+    def __init__(
+        self,
+        tools: list[BaseTool],
+        llm: BaseLanguageModel,
+        prompt: ChatPromptTemplate,
+        conversation_service: ConversationService,
+        conversation_id: int,
+    ):
+        self._tools = tools
+        self._llm = llm
+        self._prompt = prompt
+        self._conversation_service = conversation_service
+        self._conversation_id = conversation_id
+        self._memory = self._create_agent_memory(conversation_service.get_messages(self._conversation_id))
+
+    @classmethod
+    @abstractmethod
+    def create_prompt(cls, template: Sequence[MessageLikeRepresentation] | str) -> PromptTemplate | ChatPromptTemplate:
+        pass
+
+    @abstractmethod
+    def _create_agent_executor(self, **kwargs) -> AgentExecutor:
+        pass
+
+    @abstractmethod
+    def _create_agent_memory(self, messages) -> ConversationSummaryBufferMemory:
+        pass
+
+    def _set_agent_executor(self, agent_executor: AgentExecutor):
+        self._agent_executor = agent_executor
+
+    def create(self, **kwargs) -> Self:
+        agent_executor = self._create_agent_executor(**kwargs)
+        self._set_agent_executor(agent_executor)
+        return self
```

Comment:
I wonder if we should make the interface of the top-most BaseAgent this granular, or just define a constructor, and a way to query it, without even defining the implementation - any thoughts here?